xaitk_saliency/interfaces/gen_object_detector_blackbox_sal.py

```diff
@@ -0,0 +1,175 @@
+        respectively of the input reference image, (2) floating-point typed,
+        and (3) composed of values in the `[-1, 1]` range.
+
+        :param ref_image: Reference image that the input reference detections
+            belong to.
+        :param bboxes: The bounding boxes in xyxy format of the reference
+            detections to generate visual saliency maps for. This should be a
+            matrix with shape `[nDets x 4]`.
+        :param scores: The class scores of the reference detections to generate
+            visual saliency maps for. This should be a matrix with shape
+            `[nDets x nClasses]`. The order of the scores should match that
+            returned by the input black-box detection algorithm.
+        :param blackbox: The black-box object detector to perform arbitrary
+            operations on in order to deduce visual saliency.
+        :param objectness: Optional objectness score for each reference
+            detection. This should be a vector of length `nDets`. If not
+            provided, it is assumed that each detection has a score of 1.
+
+        :raises ShapeMismatchError: The implementation result visual saliency
+            heatmap matrix did not have matching height and width components to
+            the reference image.
+
+        :return: A number of visual saliency heatmaps, one for each input
+            reference detection. This is a single matrix of shape
+            `[nDets x H x W]` where `H` and `W` are the height and width
+            respectively of the input reference image.
+        """
+
+        if ref_image.ndim not in (2, 3):
+            raise ValueError(f"Input image matrix has an unexpected number of dimensions: {ref_image.ndim}")
+
+        if bboxes.shape[1] != 4:
+            raise ValueError(f"Input bounding boxes matrix has width of {bboxes.shape[1]}, should have width of 4")
+
+        # Check that reference detection inputs have matching shapes
+        if objectness is None:
+            if len(bboxes) != len(scores):
+                raise ValueError(
+                    f"Number of input bounding boxes and scores do not match: "
+                    f"(bboxes) {len(bboxes)} != {len(scores)} (scores)"
+                )
+        else:
+            if len(bboxes) != len(scores) or len(bboxes) != len(objectness):
+                raise ValueError(
+                    f"Number of input bounding boxes, scores, and objectness "
+                    f"scores do not match: (bboxes) {len(bboxes)} != "
+                    f"{len(scores)} (scores) and/or (bboxes) {len(bboxes)} != "
+                    f"{len(objectness)} (objectness)"
+                )
+
+        output = self._generate(
+            ref_image,
+            bboxes,
+            scores,
+            blackbox,
+            objectness,
+        )
+
+        # Check that the saliency heatmaps' shape matches the reference image.
+        if output.shape[1:] != ref_image.shape[:2]:
```

Comment:
We could also check that the quantity of output heatmaps matches the quantity of reference detections input.